src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java

```diff
@@ -201,6 +207,178 @@ public int hashCode() {
+            throw new NoSuchElementException();
+          }
+          int current = cursor;
+          cursor += step;
+          return current;
+        }
+      }
+
+      /** Iterator for decreasing sequences. */
+      private class DecreasingIterator extends UnmodifiableIterator<Integer> {
+        private int cursor = start;
+
+        @Override
+        public boolean hasNext() {
+          return cursor > stop;
+        }
+
+        @Override
+        public Integer next() {
+          if (!hasNext()) {
+            throw new NoSuchElementException();
+          }
+          int current = cursor;
+          cursor -= step;
+          return current;
+        }
+      }
+
+      private static int computeSize(int start, int stop, int step) {
+        final int length = Math.abs(stop - start);
+        final int absolute_step = Math.abs(step);
+        // round up (length / absolute_step) without using floats
+        return 1 + ((length - 1) / absolute_step);
+      }
+
+      private final int start;
+      private final int stop;
+      private final int step;
+      private final int size;
+
+      private RangeListView(int start, int stop, int step) {
+        this.start = start;
+        this.stop = stop;
+        this.step = step;
+        this.size = computeSize(start, stop, step);
+      }
+
+      @Override
+      public Integer get(int index) {
+        int value = start + step * index;
+        if ((step > 0 && value > stop) || (step < 0 && value < stop)) {
+          throw new ArrayIndexOutOfBoundsException(index);
+        }
+        return value;
+      }
+
+      @Override
+      public int size() {
+        return size;
+      }
+
+      /**
+       * Returns an iterator optimized for traversing range elements, since it's the most frequent
+       * operation for which ranges are used.
+       */
+      @Override
+      public Iterator<Integer> iterator() {
+        if (step > 0) {
+          return new IncreasingIterator();
+        } else {
+          return new DecreasingIterator();
+        }
+      }
+    }
+
+    private final AbstractList<Integer> contents;
+
+    private RangeList(int start, int stop, int step) {
+      this.contents = new RangeListView(start, stop, step);
+    }
+
+    @Override
+    public boolean isTuple() {
+      return false;
+    }
+
+    @Override
+    public ImmutableList<Integer> getImmutableList() {
+      return ImmutableList.copyOf(contents);
+    }
+
+    @Override
+    public SkylarkList<Integer> getSlice(Object start, Object end, Object step, Location loc,
+        Mutability mutability) throws EvalException {
+      List<Integer> sliceIndices = EvalUtils.getSliceIndices(start, end, step, this.size(), loc);
+      ImmutableList.Builder<Integer> builder = ImmutableList.builderWithExpectedSize(sliceIndices.size());
+      for (int pos : sliceIndices) {
+        builder.add(this.get(pos));
+      }
+      return MutableList.createImmutable(builder.build());
```

Comment:
Maybe implement it lazily?
https://github.com/google/skylark/blob/master/library.go#L888

Reply:
A TODO is fine for me.
Be careful with lazy slices, e.g.

```
a = [1, 2, 3]
b = a[1:]
a[2] = 5
b # [2, 3]
```